[Feature] 게시글 상세 조회 시 댓글 목록 포함 및 /조회수 관련 오류 수정

yunjeooong opened the pull request:
# [Feature] 게시글 상세 조회 시 댓글 목록 포함 및 인증/조회수 관련 오류 수정

## Description
게시글 상세 조회 시 해당 게시글의 댓글 목록을 함께 조회하는 기능을 추가했습니다.
또한, 인증되지 않은 사용자의 게시글 조회 시 발생하는 NPE와 조회수 중복 카운트 문제를 해결했습니다.

### 작업 배경
1. 게시글 상세 페이지에서 댓글 목록이 필요하여 기능 추가
2. 비로그인 사용자의 게시글 조회 시 NPE가 발생하는 문제 해결 필요
3. 동일 사용자의 반복 조회 시 조회수가 중복 증가하는 문제 해결 필요

## Changes

### 1. 댓글 조회 기능 추가
- `CommentsResponseDTO` 추가
  - 댓글 ID, 내용, 작성 시간, 작성자 정보, 소유 여부, 답글 목록 포함
  - Record 타입으로 구현하여 불변성 보장

- `CommentsRepository` 구현
  -  부모 댓글과 자식 댓글 조회 메서드 구현
  - N+1 문제 해결을 위해 fetch join 적용
  - 삭제되지 않은 댓글만 조회하도록 필터링

- `CommentsService` 구현
  - 게시글 ID로 댓글 목록 조회
  - 계층형 구조로 부모-자식 댓글 구성
  - 댓글 작성 시간 기준 정렬 (부모: 최신순, 자식: 과거순)
  
### 2. 좋아요 수 조회 기능 추가
- `PostDetailResponseDTO` 수정
  - 좋아요 수(likeCount) 필드 추가

- `PostServiceImpl` 수정
  - `getPostLikeCount` 메서드 추가하여 MongoDB에서 좋아요 수 조회
  - 게시글 상세 조회 시 좋아요 수 정보를 함께 반환하도록 개선
  
### 3. 인증 관련 오류 수정
- `PrincipalUser` 클래스 개선
  - null user 처리 로직 추가
    - 비로그인 사용자의 경우 user가 null이어도 NPE가 발생하지 않도록 처리
    - userId가 null인 경우에 대한 명시적인 처리 추가
  
  - ROLE_ANONYMOUS 권한 부여
    - 비로그인 사용자에게 기본적인 읽기 권한을 부여하기 위한 익명 권한 설정

### 3. 조회수 중복 카운트 방지 기능 개선
- `WebConfig` 수정
  - ViewCountCheckerResolver를 ArgumentResolver로 등록
  - 쿠키 기반의 조회수 중복 방지 로직이 동작하도록 설정
  ```java
  @Override
  public void addArgumentResolvers(List<HandlerMethodArgumentResolver> resolvers) {
      // ... 기존 코드 ...
      resolvers.add(viewCountCheckerResolver);  // ViewCountChecker 등록
  }
  ```

- `PostController` 수정
  ```java
  @GetMapping("/{postId}")
  public CommonResponse<PostDetailResponseDTO> findPostDetail(
      @PathVariable Long postId,
      @ViewCountChecker Boolean canAddView,  // 쿠키 기반으로 조회수 증가 여부 결정
      PrincipalUser principalUser) {
      // ...
  }
  ```

- 쿠키 기반 조회수 중복 방지 로직 구현
  - VIEWED_POSTS 쿠키를 통한 조회 이력 관리
    - 쿠키 형식: "postId1,postId2,postId3" (조회한 게시글 ID 목록)
    - 만료 시간: 24시간 (COOKIE_EXPIRE_SEC = 60 * 60 * 24)
  - ViewCountCheckerResolver에서 쿠키 검사
    - 쿠키에 해당 게시글 ID가 없으면 true 반환 (조회수 증가)
    - 쿠키에 해당 게시글 ID가 있으면 false 반환 (조회수 유지)
  - 조회 시 쿠키 자동 갱신
    - 새로운 게시글 조회 시 쿠키에 ID 추가

### 동작 방식
1. 사용자가 게시글 조회 요청
2. ViewCountCheckerResolver가 VIEWED_POSTS 쿠키 확인
3. 쿠키 존재 여부와 게시글 ID 포함 여부 확인
4. canAddView 값을 결정하여 PostController에 전달
5. PostService에서 canAddView 값에 따라 조회수 증가 여부 결정



## 테스트 결과

동일한 요청을 두번 보낸 결과 입니다. 두 번째 요청에서도  조회수가 29로 그대로 유지되었습니다.-
쿠키 기반의 조회수 중복 방지가 제대로 동작

![image](https://github.com/user-attachments/assets/dbde521d-b539-4c93-be54-d2924c33e729)
![image](https://github.com/user-attachments/assets/af636147-f79d-40d1-af08-b82d529ebdb8)

yunjeooong committed:
refactor: PostReadService를 PostServiceImpl로 통합 및 게시글 조회 로직 개선

- PostReadService의 기능을 PostServiceImpl로 통합하여 코드 중복 제거
- 게시글 상세 조회 시 비로그인 사용자도 접근 가능하도록 개선

yunjeooong committed:
feat: 게시글 상세 조회 시 댓글 목록 조회 기능 추가

- CommentsResponseDTO record 타입으로 생성
  - 댓글 ID, 내용, 작성 시간, 작성자 정보, 소유자 여부, 대댓글 목록 포함
  - 날짜 포맷 yyyy-MM-dd HH:mm:ss로 통일
  - UserDetailDto를 활용한 작성자 정보 매핑

- CommentsRepository에 댓글 조회 메서드 추가
  - 부모 댓글과 대댓글 조회를 위한 JPQL 쿼리 구현
  - fetch join을 통한 N+1 문제 해결
  - 삭제되지 않은 댓글만 조회하도록 필터링

yunjeooong committed:
feat: 게시글 댓글 조회 서비스 로직 구현

- CommentsService 인터페이스에 댓글 조회 메서드 추가
  - getPostComments(Long postId, Long currentUserId) 메서드 정의
  - 게시글 ID와 현재 사용자 ID를 기반으로 댓글 목록 조회

- CommentsServiceImpl에 댓글 조회 로직 구현
  - 부모 댓글과 대댓글을 계층 구조로 조회
  - fetch join을 통한 사용자 정보 조회 최적화
  - 삭제되지 않은 댓글만 필터링하여 반환
  - 댓글 작성 시간 기준 정렬 (부모 댓글: 내림차순, 대댓글: 오름차순)

yunjeooong committed:
fear( CommentsServiceImpl):  getPostComments메서드에 대한 주석 추가:

yunjeooong committed:
feat: 게시글 상세 조회 시 댓글 목록 포함하도록 수정

yunjeooong committed:
fix: 게시글 상세 조회 시 인증되지 않은 사용자 처리 개선

- PrincipalUser 클래스 수정
  - null user 처리 로직 추가
  - ROLE_ANONYMOUS 권한 부여 로직 추가
  - UserDetails 인터페이스의 누락된 메서드 구현
    - isAccountNonLocked()
    - isCredentialsNonExpired()

- PostController 수정
  - findPostDetail 메서드의 canAddView 파라미터 처리 개선
  - @ViewCountChecker와 @RequestParam 함께 사용하여 NPE 방지
  - defaultValue=true 설정으로 기본값 지정

결과: 인증되지 않은 사용자도 게시글을 정상적으로 조회할 수 있도록 개선

yunjeooong committed:
fix: 게시글 조회수 중복 카운트 방지 기능 개선

- PostController 수정
  - @RequestParam(defaultValue = true) 제거
  - @ViewCountChecker 어노테이션만으로 조회수 증가 여부 결정
  - 불필요한 파라미터 기본값 설정 제거

- 쿠키 기반 조회수 중복 방지 로직 정상화
  - VIEWED_POSTS 쿠키를 통한 조회 이력 관리
  - 동일 게시글 재조회 시 조회수 증가 방지
  - 쿠키 만료 시간 24시간 설정

결과: 동일한 사용자가 같은 게시글을 반복 조회해도 조회수가 중복 증가하지 않도록 개선

yunjeooong committed:
chore: 불필요한 파일 삭제

yunjeooong committed:
chore(PrincipalUser): 불필요한 메서드 정리

yunjeooong committed:
chore: CompletedDealController 파일을 develop 브랜치 상태로 복원

yunjeooong committed:
chore: 불필요한 import 문 삭제

nagul2 commented on src/main/java/com/ani/taku_backend/comments/repository/CommentsRepository.java:
```diff
@@ -2,6 +2,23 @@
+            "AND c.deletedAt IS NULL " +
+            "ORDER BY c.createdAt ASC")
+    List<Comments> findChildComments(@Param("parentId") Long parentId);
+}
```
단순히 궁금증입니다!
Query가 긴데 @Query로는 길게 쓰면 신택스 오류가 날 수 있다고 배워서.. 혹시 Querydsl안쓰시구 @Query로 사용하시는 이유가 궁금합니다~

yunjeooong replied:
첫번째로는 제가 Querydsl에 익숙하지 않은 점이 가장 큰 이유...입니당🥹
하지만 해당 쿼리들이 단순 조회 쿼리라 @Query로도 충분히 해결 가능하다고 판단했습니다!!
일반적으로 쿼리 작성 시, 단순한 기능은 JPA의 메서드 네이밍으로, 조금 복잡한 정적 쿼리는 @Query로, 동적 쿼리나 매우 복잡한 쿼리는 Querydsl로 해결하는 방식을 사용하는걸로 알고 있습니다.
 이 케이스의 경우, JOIN FETCH와 몇 가지 조건이 있는 단순 조회 쿼리로  동적 쿼리가 필요 없는 정적쿼리입니다! 이런 상황에서는  @query로도 충분히 커버가 가능하다고 생각을 했어요,,
 물론 Querydsl이 타입 안정성과 유지보수 측면에서 장점이 많아  단순한 쿼리도 Querydsl로 작성하게 되는 경우도 많다고는 알고 있지만 @Query가 더 명확하고 간결하게 요구사항을 충족할 수 있어 요 방식을 선택했습니다. 

혹시 제가 잘못 알고 있는 부분이 있으면 답변 남겨주시면 감사하겠습니다!!!! 😊

beetrbgus replied:
\+ 로 하면 나중에 유지보수가 힘들기 때문에 """ """ 을 사용해보시는건 어떨까요

nagul2 commented on src/main/java/com/ani/taku_backend/post/service/PostServiceImpl.java:
```diff
@@ -127,6 +132,37 @@ public void deletePost(Long postId, User user) {
+
+        return new PostDetailResponseDTO(post, isOwner, comments);
+    }
+
```
몽고DB에서 좋아요 수 가져오는 코드가 빠져있는것같습니다~

yunjeooong replied:
확인 감사합니다~~☺️

yunjeooong replied:
확인 감사합니다~ 추가해서 업데이트하도록 하겠습니당

yunjeooong committed:
feat: 게시글 상세 조회 시 좋아요 수 정보 추가

beetrbgus commented on src/main/java/com/ani/taku_backend/comments/service/CommentsServiceImpl.java:
```diff
@@ -135,4 +137,45 @@ private void checkAuthorAndAdmin(User user, Comments findComments) {
+    @Transactional(readOnly = true)
+    public List<CommentsResponseDTO> getPostComments(Long postId, Long currentUserId) {
+        // 최상위 댓글 조회
+        List<Comments> parentComments = commentsRepository.findParentComments(postId);
```
너무 여러번 DB를 조회 하시는 것 같습니다. 
댓글 엔티티는 대댓글과 밀접하게 연관되어 있기 때문에 
JPA의 fetch eager를 사용해도 무방할 것 같습니다.
n개의 댓글에 또 n개의 대댓글 조회를 하기 때문에 n제곱으로 효율성이 떨어질 것  같습니다.

yunjeooong replied:
규현님 의견 감사합니다. ☺️
DB를 너무 많이 접근하는 문제를 해결하기 위해 다음과 같이 개선했습니다.

1. 기존 문제점:
```java
// 최상위 댓글 조회
List<Comments> parentComments = commentsRepository.findParentComments(postId);

// 각 댓글마다 대댓글을 다시 조회 (N+1 문제 발생)
List<Comments> replies = commentsRepository.findChildComments(comment.getId());
```
- 부모 댓글 조회 1번
- 각 댓글마다 대댓글 조회 N번
- 총 N+1번의 DB 조회 발생

2. 개선된 쿼리:
```java
@Query("""
        SELECT DISTINCT c FROM Comments c
        LEFT JOIN c.parentComment p
        JOIN FETCH c.user
        WHERE c.post.id = :postId
        AND c.deletedAt IS NULL
        ORDER BY CASE WHEN c.parentComment IS NULL THEN c.createdAt ELSE p.createdAt END DESC,
                 CASE WHEN c.parentComment IS NULL THEN 0 ELSE c.createdAt END ASC
        """)
List<Comments> findParentComments(@Param("postId") Long postId);
```
- 단일 쿼리로 모든 댓글을 한 번에 조회
- LEFT JOIN으로 부모-자식 관계 처리
- JOIN FETCH로 사용자 정보 즉시 로딩
- CASE 문을 활용한 계층별 정렬
  - 부모 댓글: 최신순(DESC) 정렬
  - 자식 댓글: 시간순(ASC) 정렬

위와 같은 수정사항을 통해서 
- DB 조회 횟수: N+1 → 1회로 감소
- 불필요한 DB 호출 제거
- 필요한 데이터를 한 번에 로딩
할 수 있도록 하였습니다~

👉🏻요약
규현님 말씀대로 댓글과 대댓글은 밀접한 연관관계가 있어 함께 조회되는 것이 자연스럽고 효율적이기 때문에
LEFT JOIN과 JOIN FETCH를 활용해 한 번의 쿼리로 모든 필요한 데이터를 조회하고, CASE 문으로 계층별 정렬까지 구현하여 효율적으로 개선했습니다!

beetrbgus commented on src/main/java/com/ani/taku_backend/user/model/dto/PrincipalUser.java:
```diff
@@ -32,14 +32,18 @@ public User getUser() {
     }
 
     public Long getUserId() {
```
이분에서 삼항연산자를 쓴이유가 궁금합니다.
user가 null이 아니어도 user.getUserId가 Long 클래스이기 때문에 null일 가능성이 있어서 NPE를 예방한다라고 볼 수는 없을 것 같습니다.

yunjeooong replied:
규현님 말씀대로 `userId`가 Long 타입이라 null 가능성이 있는 것은 맞습니다!
하지만  삼항 연산자를 사용한 이유는 user.getUserId()의 반환값이 null일 수 있어서가 아니라, 
비로그인 사용자의 경우 user 객체 자체가 null이기 때문입니다. (실제로 NPE문제가 발생)

두 가지 상황을 구분해보면:

1. user 객체가 null인 경우:
   - 비로그인 사용자일 때 발생
   - 이때 user.getUserId() 호출 시 NPE 발생
   - 삼항 연산자로 이 상황 방지

2. user.getUserId()가 null인 경우:
   - 리뷰어님이 지적하신 것처럼 Long 타입이라 null 가능
   - 하지만 이 경우는 NPE가 발생하지 않음
   - 단순히 null 값이 반환되는 상황

하지만 다시 생각해보니  PrincipalUser는 인증된 사용자의 정보만을 다루도록 하는 class인데 비로그인 책임을 주는게 어색하다고 느껴서 기존 PrincipalUser를 유지하고 서비스 레이어에서 명시적인 null 체크를 통해 비로그인 사용자를 처리하는 방식으로 개선했습니다. 

1. 이전 접근 방식:
```java
// PrincipalUser.java
public Long getUserId() {
    return user != null ? user.getUserId() : null;  // user 객체의 null 여부를 체크
}

// PostServiceImpl.java
boolean isOwner = false;
if (post.getUser() != null && currentUserId != null &&
        post.getUser().getUserId().equals(currentUserId)) {
    isOwner = true;
}
```

2. 개선된 접근 방식:
```java
// PrincipalUser.java - 기존 코드 유지
public Long getUserId() {
    return user.getUserId();
}

// PostServiceImpl.java - 비로그인 사용자 처리 로직 이동
boolean isOwner = false;
if (currentUserId != null && post.getUser() != null) {  // 명시적인 null 체크
    isOwner = post.getUser().getUserId().equals(currentUserId);
}
```

답변주셔서 정말 감사합니다👍👍👍

beetrbgus commented on src/main/java/com/ani/taku_backend/user/model/dto/PrincipalUser.java:
```diff
@@ -52,7 +56,7 @@ public String getPassword() {
     @Override
     public String getUsername() {
-        return user.getEmail();
+        return user != null ? user.getEmail() : null;
```
이것들도 마찬가지 인 것 같습니다.

yunjeooong committed:
Merge branch 'develop' into feature/250126_postdetails

yunjeooong committed:
style: CommentsRepository의 JPQL 쿼리를 텍스트 블록으로 변경

yunjeooong committed:
refactor: PrincipalUser와 PostServiceImpl의 책임 분리

- PrincipalUser에서 비로그인 사용자 처리 로직 제거
  - 삼항 연산자를 사용한 null 체크 제거
  - 인증된 사용자 정보만을 다루도록 수정

- PostServiceImpl에서 비로그인 사용자 처리 로직 추가
  - 명시적인 null 체크를 통한 비로그인 사용자 처리
  - isOwner 판단 로직 개선

yunjeooong committed:
fix: 댓글 조회 시 N+1 문제 해결

- 댓글 조회 쿼리 최적화
  - 단일 쿼리로 부모 댓글과 자식 댓글을 한 번에 조회하도록 개선
  - LEFT JOIN을 사용하여 부모-자식 댓글 관계 처리
  - JOIN FETCH로 댓글 작성자 정보 즉시 로딩

- Repository 변경사항
  - findParentComments 쿼리 개선
    - 기존: 부모 댓글 조회 후 각 자식 댓글 개별 조회 (N+1 발생)
    - 개선: LEFT JOIN으로 한 번에 모든 댓글 조회

- 성능 개선 효과
  - DB 조회 횟수: N+1 → 1회로 감소
  - 불필요한 DB 호출 제거

yunjeooong committed:
fix:정렬 로직 개선

yunjeooong committed:
병합 충돌 해결Merge branch 'feature/250126_postdetails' of https://github.com/taku-project/taku-backend into feature/250126_postdetails
